util/generate-rsa-signing-tests.py

```diff
@@ -40,57 +51,161 @@ def parse(fn):
+    "# Digest = SHAAlg.\n" + \
+    "# Key is a DER-formatted PKCS1-encoded RSA key.\n" + \
+    "# Sig = S.\n\n"
+
+    num_cases = 0
+
+    # Each test type has a different field as the last entry per case
+    if test_type == "verify":
+        last_field = "Result"
+    elif padding_alg == "PSS":
+        last_field = "SaltVal"
+    else:
+        last_field = "S"
+    for case in parse(fn, last_field):
 
         if case['SHAAlg'] == 'SHA224':
             # SHA224 not supported in *ring*.
+            if DEBUG:
+                print "Skipping due to use of SHA224"
+            continue
+
+        if case['SHAAlg'] == 'SHA1' and padding_alg == 'PSS':
+            # SHA1 with PSS not supported in *ring.
+            if DEBUG:
+                print "Skipping due to use of SHA1 and PSS."
             continue
 
+        # We only support PSS where the salt length is equal to the output
+        # length of the hash algorithm
+        if padding_alg == "PSS":
+            if len(case['SaltVal']) * 2 != digest_output_lens[case['SHAAlg']]:
+                if DEBUG:
+                    print "Skipping due to incorrect salt length."
+                continue
+
         # Read private key components.
         n = int(case['n'], 16)
         e = int(case['e'], 16)
         d = int(case['d'], 16)
 
-        # Recover the prime factors and CRT numbers.
-        p, q = rsa.rsa_recover_prime_factors(n, e, d)
-        # cryptography returns p, q with p < q by default. *ring* requires
-        # p > q, so swap them here.
-        p, q = max(p, q), min(p, q)
-        dmp1 = rsa.rsa_crt_dmp1(d, p)
-        dmq1 = rsa.rsa_crt_dmq1(d, q)
-        iqmp = rsa.rsa_crt_iqmp(p, q)
+        if n.bit_length() < 2048:
+            if DEBUG:
+                print "Skipping due to modulus length (too small)."
+            continue
+
+        if n.bit_length() > 4096:
+            if DEBUG:
+                print "Skipping due to modulus length (too large)."
+            continue
+
+        # Can only support moduli which are a whole number of bytes
+        if n.bit_length() % 8 != 0:
+            if DEBUG:
+                print "Skipping due to modulus (not a whole number of bytes)."
+            continue
+
+        if test_type == 'sign':
+            # Recover the prime factors and CRT numbers.
+            p, q = rsa.rsa_recover_prime_factors(n, e, d)
+            # cryptography returns p, q with p < q by default. *ring* requires
+            # p > q, so swap them here.
+            p, q = max(p, q), min(p, q)
+            dmp1 = rsa.rsa_crt_dmp1(d, p)
+            dmq1 = rsa.rsa_crt_dmq1(d, q)
+            iqmp = rsa.rsa_crt_iqmp(p, q)
 
         # Create a private key instance.
         pub = rsa.RSAPublicNumbers(e, n)
-        priv = rsa.RSAPrivateNumbers(p, q, d, dmp1, dmq1, iqmp, pub)
-        key = priv.private_key(default_backend())
 
-        # Recalculate and compare the signature to validate our processing.
+        if test_type == 'sign':
```

Comment:
Since this is first of three subsequent `if test_type == 'sign'`s, I would suggest it's probably better to factor out the different types into different functions, e.g. have `print_sign_test()` and `print_verify_test()` that you can then dispatch to. You'll probably then want a helper function that has some of the `print` statements that are common to both.